docs/proposals/20210316-containerrecreaterequest.md

```diff
@@ -0,0 +1,108 @@
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+Now that the only way to recreate a container in an active Pod is to update it in-place.
+However, it also requires users to change the image defined in workloads (e.g. CloneSet/Advanced StatefulSet).
+
+We hope to provide a way to let users recreate(restart) one or more containers in any active Pods.
+It is useful for some user scenarios, such as:
+
+1. Restart the application in app container.
+2. Restart a sidecar contaienr along with app container.
+
+## Proposal
+
+Define a CRD named `ContainerRecreateRequest` (short name `CRR`):
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: ContainerRecreateRequest
+metadata:
+  namespace: pod-namespace
+  name: xxx
+  labels:
+    crr.apps.kruise.io/pod-name: pod-name
+    crr.apps.kruise.io/node-name: node-name-of-the-pod  # injected by kruise webhook
+spec:
+  containers:
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'
+    stopGracePeriodSeconds: 30   # the timeout period for preStop hook and container stopping
+    readyProtection: false       # 'true' means make the Pod not-ready before recreating containers
+    continueByReady: false       # 'true' means recreate the next container only if the previous one has been recreated and ready
+  activeDeadlineSeconds: 300        # deadline of this ContainerRecreateRequest
+  ttlSecondsAfterFinished: 1800     # ttl after this ContainerRecreateRequest has finished
+```
+
+1. The containers will be recreated one by one.
+   Typically, we will stop the next one once if the previous has stopped, unless the `continueByReady` is `true`.
+2. The `readyProtection` feature depends on a new feature-gate named `PodRecreateReadinessGate`,
+   which indicates to inject a readinessGate during each Pod creating.
+
+### User Stories
+
+#### Story 1
+
+To restart the service running in one container, user needs to restart the specific container.
+It is too EXPENSIVE to recreate the whole Pod.
+
+Related [Issue](https://github.com/openkruise/kruise/issues/495).
+
+### Requirements
+
+`ContainerRecreateRequest` relies on kruise-daemon to stop the container in Pod.
+
+So if the `KruiseDaemon` feature-gate is closed, `ContainerRecreateRequest` will also be disabled.
+
+### Implementation Details/Notes/Constraints
+
+![container_recreate](../../docs/img/containerrecreate.png)
+
+For containers, **kruise-daemon** will stop the container with or without preStop hook,
+then **kubelet** will create a new container and start again.
+
+While some containers being recreating, the other containers in the same Pod will still be running.
+
+### Risks and Mitigations
+
+- Recreating a container will cause the application in it to start again.
```

Comment:
is it possible that the recreation and inplace-update of one pod occur in the same time? please describe the possible problem and solutions.